src/coreclr/vm/jitinterface.cpp

```diff
@@ -11798,7 +11798,16 @@ InfoAccessType CEEJitInfo::constructStringLiteral(CORINFO_MODULE_HANDLE scopeHnd
     }
     else
     {
-        *ppValue = (LPVOID)ConstructStringLiteral(scopeHnd, metaTok); // throws
+        void*** ptr = reinterpret_cast<void***>(ConstructStringLiteral(scopeHnd, metaTok));
+        if ((ptr != nullptr) && (*ptr == nullptr))
```

Comment:
What is this check for?

Reply:
@jkotas not sure yet but it fixed Access violation-like failures for me, will investigate.

Reply:
I had to make a few changes in the JIT because it doesn't expect any kind of constants of `TYP_REF` other than `null` at the moment, it seems there was some support for raw string handles for CoreRT (Referenced as "frozen" - I guess some sort of embedded into the data section?) but still hit a couple of asserts. I'm glad the CI is now green.

Reply:
@jkotas so that check indeed was redundant - I was just hitting a random gc hole, but turns out some strings (not literals) can get into the same cache from some other places (I suspect via interning) and be on anything other than POH.
I added a hack-check "ObjGeneration == 2" (it's not enough to make sure it's in POH but still) and it fixed the gc hole problem for me for tests.

Reply:
Repro:
```csharp
using System;
using System.Runtime.CompilerServices;

class Pr
{
    static void Main()
    {
        string.Intern(42.ToString());
        Test1();
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void Test1() => 
        Console.WriteLine(GC.GetGeneration("42")); // prints 0 in my branch
}
```

Reply:
Ok, so it is the same problem as https://github.com/dotnet/runtime/pull/49576#discussion_r594396313

Reply:
oops, indeed

Reply:
Does it mean we should move users' string objects to POH if they call string.intern over them?
Maybe it's better to introduce a gchelper method "is in poh" and leave interned strings as is?

Reply:
`string.intern` is frowned upon. I do not see a problem with making it slower.

> Maybe it's better to introduce a gchelper method "is in poh" and leave interned strings as is?

I would lead to unpredictable performance. If somebody calls string.intern, JITed code using the same string will get slower. I do not think it is a good tradeoff.